### ISSUE TITLE
Python pitch and onset detection from np.array?

### ISSUE THREAD

**Issue by asdfang** [aubio/aubio#105]
I found an example on how to use pitch detection from a numpy array in demo_pitch_sinusoid.py.
demo_pitch.py and demo_onset.py both read a file from the command line, so I was happy to find demo_pitch_sinusoid.py as an example to detect pitches from a numpy array.

Is there a way to do this for the onset detector as well?

Thanks

**Comment by piem** [comment]
hi @asdfang 

Sure, you simply need to feed the onset algorithm with blocks of size `hop_size`, same as what is done in `demo_pitch_sinusoid.py`. Using `.reshape(-1, hop_size)` on your input vector (padded with zeros so that the total length is a multiple of `hop_size`) should be enough. Another example you might want to look at is `demo_bench_yin.py`.

Let us know how this goes.

**Comment by asdfang** [comment]
Hi,

I get this error when I try to run the demo_bench_yin.py demo:

I would prefer to use in the miniconda environment, how do I get rid of the
one installed to the system?
Sorry, I'm new to managing packages....
Is there a different problem to this? I ask because I've used aubio with
those warnings and it still compiled fine.

Thanks,
Alex

**Comment by piem** [comment]
hi @asdfang 

it seems you forgot to paste the error. if you installed aubio with pip, you should be able to run `pip uninstall aubio` to uninstall previous version.

let us know how it goes!

**Comment by asdfang** [comment]
Hi,

I pasted a screenshot inside the email, but perhaps it doesn't work when it
transfers to github. Here is a text version:

(HarmonizeMe)dhcp-10-105-178-161:python_tests asdfang$ python
demo_bench_yin.py

objc[97987]: Class TKApplication is implemented in both
/Users/asdfang/miniconda2/envs/HarmonizeMe/lib/libtk8.5.dylib and
/System/Library/Frameworks/Tk.framework/Versions/8.5/Tk. One of the two
will be used. Which one is undefined.

objc[97987]: Class TKMenu is implemented in both
/Users/asdfang/miniconda2/envs/HarmonizeMe/lib/libtk8.5.dylib and
/System/Library/Frameworks/Tk.framework/Versions/8.5/Tk. One of the two
will be used. Which one is undefined.

objc[97987]: Class TKContentView is implemented in both
/Users/asdfang/miniconda2/envs/HarmonizeMe/lib/libtk8.5.dylib and
/System/Library/Frameworks/Tk.framework/Versions/8.5/Tk. One of the two
will be used. Which one is undefined.

objc[97987]: Class TKWindow is implemented in both
/Users/asdfang/miniconda2/envs/HarmonizeMe/lib/libtk8.5.dylib and
/System/Library/Frameworks/Tk.framework/Versions/8.5/Tk. One of the two
will be used. Which one is undefined.

2017-04-28 14:48:36.008 python[97987:15991547] setCanCycle: is deprecated.
Please use setCollectionBehavior instead

2017-04-28 14:48:36.040 python[97987:15991547] setCanCycle: is deprecated.
Please use setCollectionBehavior instead

Exception in Tkinter callback

Traceback (most recent call last):

  File
"/Users/asdfang/miniconda2/envs/HarmonizeMe/lib/python2.7/lib-tk/Tkinter.py",
line 1536, in __call__

    return self.func(*args)

  File
"/Users/asdfang/miniconda2/envs/HarmonizeMe/lib/python2.7/site-packages/matplotlib/backends/backend_tkagg.py",
line 283, in resize

    self.show()

  File
"/Users/asdfang/miniconda2/envs/HarmonizeMe/lib/python2.7/site-packages/matplotlib/backends/backend_tkagg.py",
line 355, in draw

    tkagg.blit(self._tkphoto, self.renderer._renderer, colormode=2)

  File
"/Users/asdfang/miniconda2/envs/HarmonizeMe/lib/python2.7/site-packages/matplotlib/backends/tkagg.py",
line 30, in blit

    id(data), colormode, id(bbox_array))

TclError

*********************************

When I don't run it in the miniconda environment, aubio is not installed.
Ideas?


Thanks for getting back so quickly,

Alex

**Comment by piem** [comment]
@asdfang this seems to be an installation issue of your `matplotlib` package, not `aubio`. Looking for this error message [leads to answers](http://stackoverflow.com/a/37355350/2118678)  mentioning that uninstalling the `matplotlib` package installed in your `/System` is one of the way to go. Another answer mentions [defining a custom matplotlib backend](http://stackoverflow.com/a/34818746/2118678).

Hope this helps! Let us know when you get it to work.

**Comment by asdfang** [comment]
Great! I sudo pip uninstall-ed matplotlib and reinstalled it in miniconda
environment (after trying and failing to individually remove it from
different pythons on my machine....) and now it works.

As for using this to see how to set up the onset algorithm...I'm not
connecting the dots just yet.

I took the code in demo_pitch_sinusoid.py and just tried to transfer it to
onsets and came up with:

def run_onset(o, input_vec):
onsets = []
for vec_slice in input_vec.reshape((-1, o.hop_size)):
a = o(vec_slice)[0]
onsets.append(a)
return onsets

def aubio_onsets(audio):
a = audio
onsets = []
o = onset("default", frameSize, hopSize, sampleRate)
onsets = run_onset(o, a)
return onsets

It returns to me a list (whose length is equal to the number of frames)
with mostly 0.0's, and sporadic nonzero elements.
I collected the indices at which there were nonzero elements, and
multiplied it by hopSize to get it in terms of samples,
and received numbers that are within hopSize of what returns if I ran the
normal onset detector with the file path and not through just a float array.
So, it seems it's alright! I'm sure this is the effect of what that while
True loop is doing in demo_onsets.py (but that code is going over my head).
Now I'm wondering how the onset algorithm noted an onset in between
hopSizes?


Thank you so much for being so responsive and helpful, let me know if
anything above looks wrong to you!
Alex
